onnxruntime/contrib_ops/cpu/ngram.cc

```diff
@@ -0,0 +1,519 @@
+    ORT_ENFORCE(status.IsOK() && !pool_int64s.empty(), "non-empty pool_int64s is required if pool_strings not provided");
+  }
+
+  // Iterator via the pool. Insert 1 item for 1-grams, 2 items for 2-grams, etc.
+  const auto total_items = (impl_->pool_strings_.empty()) ? pool_int64s.size() : impl_->pool_strings_.size();
+  size_t ngram_id = 0;
+  // Load into dictionary only required Ns
+  const size_t M = impl_->M_;
+  const size_t N = impl_->N_;
+  size_t ngram_size = 1;
+  for (size_t i = 0; i < impl_->ngram_counts_.size(); ++i) {
+    size_t start_idx = impl_->ngram_counts_[i];
+    size_t end_idx = ((i + 1) < impl_->ngram_counts_.size()) ? impl_->ngram_counts_[i + 1] : total_items;
+    ORT_ENFORCE(end_idx >= start_idx && end_idx <= total_items,
+                "n-gram counts out of bounds for ", std::to_string(ngram_size), "-grams");
+    auto items = end_idx - start_idx;
+    if (items > 0) {
+      ORT_ENFORCE((items % ngram_size == 0),
+                  "Number of items must compose whole ", std::to_string(ngram_size), "-grams");
+      auto ngrams = items / ngram_size;
+      // Skip loading into hash_set ngrams that are not N or not in the range of [M-N] for all=true;
+      if (ngram_size >= M && ngram_size <= N) {
+        if (impl_->pool_strings_.empty()) {
+          auto before_insert = impl_->int_set_.size();
+          Emplace(pool_int64s.begin() + start_idx, ngrams, ngram_size, ngram_id, impl_->int_set_);
+          ORT_ENFORCE((before_insert + ngrams) == impl_->int_set_.size(), "pool_int64s duplicate ", std::to_string(ngram_size), "-grams detected");
+        } else {
+          auto before_insert = impl_->str_set_.size();
+          Emplace(impl_->pool_strings_.begin() + start_idx, ngrams, ngram_size, ngram_id, impl_->str_set_);
+          ORT_ENFORCE((before_insert + ngrams) == impl_->str_set_.size(), "poll_strings duplicate ", std::to_string(ngram_size), "-grams detected");
+        }
+      } else {
+        ngram_id += ngrams;
+      }
+    }
+    ++ngram_size;
+  }
+}
+
+Ngram::~Ngram() {
+}
+
+void Ngram::OutputResult(OpKernelContext* ctx, const std::vector<uint32_t>& frequences) const {
+  std::vector<int64_t> output_dims;
+  output_dims.push_back(frequences.size());
+
+  TensorShape output_shape(output_dims);
+  auto Y = ctx->Output(0, output_shape);
+  auto output_data = Y->MutableData<float>();
+  const auto& w = impl_->weights_;
+  switch (impl_->weighting_criteria_) {
+    case kTF: {
+      for (auto f : frequences) {
+        *output_data++ = static_cast<float>(f);
+      }
+    } break;
+    case kIDF: {
+      if (!w.empty()) {
+        assert(frequences.size() == w.size());
+        for (size_t i = 0; i < frequences.size(); ++i) {
+          *output_data++ = (frequences[i] > 0) ? w[i] : 0;
+        }
+      } else {
+        for (auto f : frequences) {
+          *output_data++ = (f > 0) ? 1.0f : 0;
+        }
+      }
+      break;
+      case kTFIDF: {
+        if (!w.empty()) {
```

Comment:
Indent is off

Reply:
I have verified that '}' is not missing. The indent is not off, it shows all right in the Studio where I have whitespace shown. I also have an automatic clang formatting on. 

---
In reply to: [243163037](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243163037) [](ancestors = 243163037)